refactor: refactor structure to: shareit-server, shareit-gateway   feat: add gateway for All models   feat: add docker-compose for ShareIt_Gateway, ShareIt_Server, Shareit_DB

RodionovRL opened the pull request.

RodionovRL committed:
refactor: refactor structure to: shareit-server, shareit-gateway
feat: add gateway for Users

RodionovRL committed:
feat: add gateway for All models
feat: add docker-compose for ShareIt_Gateway, ShareIt_Server, Shareit_DB

kvach2avl commented on docker-compose.yml:
```diff
@@ -0,0 +1,39 @@
+      - server
+    environment:
+      - SHAREIT_SERVER_URL=http://server:9090
+#      - TZ=Europe/Moscow
```
Здесь и ниже: так же, как и в java-коде, не стоит оставлять в конфигах закомментированных строк (если это не пояснения к написанному, конечно)

kvach2avl commented on gateway/Dockerfile:
```diff
@@ -0,0 +1,4 @@
+FROM amazoncorretto:11
+ENV JAVA_TOOL_OPTIONS -agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=*:8081
```
Настройки для отладки надо убрать: для отладки локально это допустимо, но в "боевых" конфигах такого лучше не делать. Да и не даст никто отлаживаться в проде

kvach2avl commented on gateway/src/main/java/ru/practicum/shareit/exception/ErrorHandler.java:
```diff
@@ -0,0 +1,22 @@
+
+@Slf4j
+@RestControllerAdvice
+public class ErrorHandler {
```
Здесь не хватает обработчика исключений, возникающих при взаимодействии с сервером. Необходимо добавить обработку Thowable с генерированием http/500

RodionovRL replied:
Любопытный момент. Подскажите, пожалуйста, а как обрабатывать только необработанные исключения. Ибо, если я добавляю обработку Trowable, то у меня все исключения обрабатываются как  http/500.

RodionovRL replied:
Кажется разобрался =)

kvach2avl commented on server/pom.xml:
```diff
@@ -0,0 +1,134 @@
+            <artifactId>spring-boot-starter-test</artifactId>
+            <scope>test</scope>
+        </dependency>
+<!--        <dependency>-->
```
Долой комментированные настройки

kvach2avl commented on server/src/main/java/ru/practicum/shareit/booking/dto/BookItemRequestDtoT.java:
```diff
@@ -0,0 +1,20 @@
+@AllArgsConstructor
+public class BookItemRequestDtoT {
+	private long itemId;
+	@FutureOrPresent
```
Поскольку все проверки пользовательского ввода вынесены на сторону gateway, то и соответствующие аннотации в серверных компонентах можно удалить

kvach2avl commented on server/src/main/java/ru/practicum/shareit/booking/dto/BookingInputDto.java:
```diff
@@ -0,0 +1,26 @@
+@NoArgsConstructor
+public class BookingInputDto {
+    @NotNull
+    @FutureOrPresent(message = "Start booking may be only in present or future")
```
Аналогично...

kvach2avl reviewed (commented):
В целом неплохо, но кое-что надо доработать. В основном это касается проверок на стороне сервера - там должны остаться только бизнес-проверки, те, что связаны с логикой работы приложения. Всё, что касается корректности ввода данных (в т.ч. соответствующие аннотации), должно делаться в gateway

RodionovRL committed:
fix: remove needless validation at the server
fix: remove unnecessary code-comments
fix: add handle InternalServerError HTTP/500

kvach2avl commented on gateway/src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -0,0 +1,81 @@
+                                             @RequestBody @Valid BookingInputDto requestDto
+    ) {
+        log.info("GW Creating booking {}, userId={}", requestDto, userId);
+        return bookingClient.addBooking(userId, requestDto);
```
По хорошему, сюда стоит перенести проверку, выполняемую нынче в BookingServiceImpl.addBooking() методом validateBookingData(). Будет обидно протащить через gateway в server некорректный запрос на создание бронирования только ради того, чтобы потом обработать в gateway-е серверное исключение

kvach2avl commented on server/pom.xml:
```diff
@@ -0,0 +1,126 @@
+
+        <dependency>
+            <groupId>jakarta.validation</groupId>
+            <artifactId>jakarta.validation-api</artifactId>
```
на сервере эта зависимость лишняя, все проверки корректности ввода по ТЗ делаются на стороне gateway

kvach2avl commented on server/src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -6,8 +6,8 @@
+import ru.practicum.shareit.booking.dto.BookingInputDto;
 import ru.practicum.shareit.booking.service.api.BookingService;
 
 import javax.validation.Valid;
```
здесь и в других классах сервера: эту и пару следующих аннотаций нужно убрать. Все проверки корректности ввода о ТЗ делаются на стороне gateway

kvach2avl reviewed (commented):
часть замечаний осталась в силе

RodionovRL committed:
fix: remove needless validation at the server
refactor: move validateBookingData from server to gateway

RodionovRL committed:
fix: fix codestyle

kvach2avl reviewed (commented):
Ok
